plugins/storage/object/cloudian/src/main/java/org/apache/cloudstack/storage/datastore/driver/CloudianHyperStoreObjectStoreDriverImpl.java

```diff
@@ -0,0 +1,877 @@
+
+        // persist the root and iam credentials in the database and update all bucket details.
+        persistCredentials(storeId, accountId, details, credential, iamCredential);
+        return true;
+    }
+
+    /**
+     * Create IAM credentials if required.
+     *
+     * When the HyperStore user is first created, this method will create an IAM User with an appropriate
+     * permission policy and a set of credentials which will be returned.
+     * After the first run, the IAM resources should already be in place in which case we just ensure
+     * the credentials we are using are still available and if not, it tries to recreate the IAM resources.
+     *
+     * @param storeId the store
+     * @param details a map of existing account details that we know about including any saved IAM credentials.
+     * @param credential the account Root User credentials (to manage the IAM resources).
+     * @return an AccessKey object for newly created IAM credentials or null if existing credentials were ok
+     *    and nothing was created.
+     */
+    protected AccessKey createIAMCredentials(long storeId, Map<String, String> details, CloudianCredential credential) {
+        AmazonIdentityManagement iamClient = getIAMClientByStoreId(storeId, credential);
+        final String iamUser = CloudianHyperStoreUtil.IAM_USER_USERNAME;
+
+        // If an accessKeyId is known to us, check IAM still has it.
+        String iamAccessKeyId = details.get(CloudianHyperStoreUtil.KEY_IAM_ACCESS_KEY);
+        if (iamAccessKeyId != null) {
+            try {
+                logger.debug("Looking for IAM credential {} for IAM User {}", iamAccessKeyId, iamUser);
+                ListAccessKeysResult listAccessKeyResult = iamClient.listAccessKeys(new ListAccessKeysRequest().withUserName(iamUser));
+                for (AccessKeyMetadata accessKeyMetadata : listAccessKeyResult.getAccessKeyMetadata()) {
+                    if (iamAccessKeyId.equals(accessKeyMetadata.getAccessKeyId())) {
+                        return null;  // The IAM AccessKeyId still exists (as expected). return null.
+                    }
+                    // Usually, there will only be 1 credential that we manage, but an error persisting
+                    // credentials might leave an un-managed credential which we can just delete. It is better
+                    // to delete as otherwise, we may hit a max credential limit for this IAM user.
+                    DeleteAccessKeyRequest deleteAccessKeyRequest = new DeleteAccessKeyRequest();
+                    deleteAccessKeyRequest.setUserName(iamUser);
+                    deleteAccessKeyRequest.setAccessKeyId(accessKeyMetadata.getAccessKeyId());
+                    logger.info("Deleting un-managed IAM AccessKeyId {} for IAM User {}", accessKeyMetadata.getAccessKeyId(), iamUser);
+                    iamClient.deleteAccessKey(deleteAccessKeyRequest);
```

Comment:
Is it really ok to delete the credential in this method? If some error created a lot of unmanaged credentials, shouldn't the admin delete them manually? (asking as a layman on Cloudian)

Reply:
Hi @JoaoJandre . Thanks for asking. Indeed it's a little strange at first but I think it is the right thing to do. This IAM user is owned by cloudstack. There should be nothing else creating IAM credentials for this user except for this plugin. IAM resources such as user/credentials/policies etc are the responsibility of the account so the administrator would not normally get involved with managing them. Also the account itself is owned by cloudstack and the idea is that you can only create buckets via the cloudstack interface and the credentials you have access to as a user also have no IAM access. That's basically why in this particular case, I think it is best to tidy up directly. Hope that answers your question.

Reply:
Ok, that makes sense then. In any case, this user deletion code is duplicated, we could extract it to a method.

Reply:
Done! Thanks for the suggestion.